plutus-core/untyped-plutus-core/test/DeBruijn/FlatNatWord.hs

```diff
@@ -0,0 +1,51 @@
+module DeBruijn.FlatNatWord (test_flatNatWord) where
+
+import Common
+import Flat
+import Hedgehog
+import Hedgehog.Gen qualified as Gen
+import Hedgehog.Range qualified as Range
+import Numeric.Natural
+import Test.Tasty.Hedgehog
+
+{- NOTE: [DeBruijjn index flat format]
+In the debruijn branch we are transitioning the debruijn index to be a Word instead of Natural, for CEK performance reasons.
+
+This means the types of `Script`,`Program`,`Term` change and so does the flat "interface".
+
+This module tests that although the interface/types change,
+the underlying flat encoding/decoding remains the same,
+so there is no need to issue a new plutus language version increase.
+-}
+
+test_flatNatWord :: TestNested
+test_flatNatWord = testNested "FlatNatWord"
+    [ pure $ testProperty "natEncByteString = wordEncByteString" prop_Enc
+    , pure $ testProperty "roundtripWord" prop_TripWord
+    , pure $ testProperty "roundtripNat" prop_TripNat
+    ]
+
+prop_Enc :: Property
+prop_Enc = property $ do
+    w <- forAll $ Gen.word64 Range.linearBounded
```

Comment:
Ah, a relevant difference: `Natural` is arbitrary-precision, `Word` is fixed-precision. Not sure if that matters :thinking:

Reply:
It only matters if someone could have used an index of > 2^64 *and had it pass scope-checking*, which would mean they would actually have had to have 2^64 lambdas, which wouldn't fit. So it's okay.

Reply:
y indeed. there is no way to go around this, but as we discussed , a script reaching a debruijn index of `maxBound:: Word64` would not fit on the chain

Reply:
I was talking to Nikos about scope checking earlier and wondered if we could safely omit it in a later version of Plutus: if someone submits a script with an out-of-scope variable then it should fail harmlessly and it'll be their own fault.  I wonder if this is a reason not to do that?  I'll think again tomorrow.

Reply:
@kwxm the issue is programs that don't look at the out-of-scope variable. For example: `\x . y`. At the moment the CEK machine accepts this, returning the term (with it's uninspected free variables intact!).

BUT the current pipeline in `plutus-ledger-api` un-debruijnifies first, which rejects free variables *anywhere*. So we had to retain that behaviour, otherwise some scripts that didn't succeed before might start succeeding.

We *could* drop this behaviour for V2, though.

Reply:
> At the moment the CEK machine accepts this, returning the term (with it's uninspected free variables intact!).

Yep, but I think that would be acceptable.  The machine also accepts things like `[(force (builtin ifThenElse)) (con integer 5) (con string "Oh no!)], which could lead to programs that succeed with some inputs but fail at run-time with a type error with other inputs.  That's not going to change, and the case of unused out-of-scope variables seems very similar.  I appreciate that we have to 
preserve the current semantics for V1 though.

Reply:
Oh yeah, I think it's totally acceptable. The thing that's annoying is that we currently effectively force the strict behaviour when we don't need to.